src/index.js

```diff
@@ -349,131 +278,236 @@ export default class RestClient {
+   * @return {Object}
+   */
+  _dedupFetch(path, tracker, resource, fetchID) {
+    const index = tracker.active.findIndex(obj => obj.value === resource);
+    const activeMatch = (index !== -1) ? tracker.active.splice(index, 1)[0] : null;
+    const activeFetchID = activeMatch ? activeMatch.fetchID : fetchID;
+    const fetchMatch = tracker.fetching.find(obj => obj.value === resource);
+    const promises = [];
+
+    if (!fetchMatch) {
+      tracker.fetching.push({ fetchID: activeFetchID, value: resource });
+      return { promises };
+    }
+
+    if (!tracker.pending[resource]) tracker.pending[resource] = [];
+
+    promises.push(this._setPendingRequest(
+      tracker.pending[resource], fetchMatch.fetchID, activeFetchID,
+    ));
+
+    this._updatePendingRequests(path, resource, fetchMatch.fetchID);
+    return { pending: true, promises };
+  }
+
+  /**
+   *
+   * @private
    * @param {string} endpoint
    * @param {Array<string>} values
    * @param {Object} context
    * @param {Object} options
-   * @param {string} resourceKey
    * @return {Promise}
    */
-  async _delete(endpoint, values, context, options, resourceKey) {
-    const res = await this._fetch('DELETE', endpoint, context, options);
-    if (res.errors) return res.errors;
+  async _delete(endpoint, values, context, options) {
+    const method = 'DELETE';
+    const fetchOptions = { headers: new Headers(options.headers), method };
 
-    if (res.data) {
-      const data = castArray(res.data);
-      this._deleteCache(data, context.path, context.queryParams, resourceKey);
-    }
+    const { data } = await this._fetch(
+      method, endpoint, fetchOptions, context, options, { resource: values },
+    );
 
-    return res.data;
+    return data;
   }
 
   /**
    *
    * @private
-   * @param {Array<Object>} data
-   * @param {string} path
-   * @param {Object} queryParams
-   * @param {string} key
+   * @param {string} endpoint
    * @return {Promise}
    */
-  async _deleteCache(data, path, queryParams, key) {
+  async _deleteCacheEntry(endpoint) {
     if (this._disableCaching) return;
 
-    data.forEach((value) => {
-      if (!get(value, [key], null)) return;
-      const endpoints = this._buildEndpoints({ path, queryParams, resource: [value[key]] });
-      if (!get(endpoints, [0, 'endpoint'], null)) return;
-
-      try {
-        this._cache.delete(endpoints[0].endpoint);
-      } catch (err) {
-        logger.error(err);
-      }
-    });
+    try {
+      this._cache.delete(endpoint);
+    } catch (err) {
+      logger.error(err);
+    }
   }
 
   /**
    *
    * @private
    * @param {string} method
    * @param {string} endpoint
+   * @param {Object} fetchOptions
    * @param {Object} context
    * @param {Object} options
-   * @param {any} content
+   * @param {Object} [metadata]
    * @return {Promise}
    */
-  async _fetch(method, endpoint, context, options, content) {
+  async _fetch(method, endpoint, fetchOptions, context, options, metadata = {}) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 19 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS3776)